### ISSUE TITLE
Request: Strict parsing mode

### ISSUE THREAD

**Issue by nzakas** [csstree/csstree#301]
I know the design of CSSTree is intended to mimic how the browser CSS parser works, so it's fault-tolerant.

I'm looking at using CSSTree for linting CSS, and this fault-tolerance means it's difficult to catch a lot of errors that might occur. As an example, `"a {` parses correctly and does not call `onParseError()` at all, so I have no way of letting folks know that they forgot to close their block.

So this is actually a couple of requests:

1. An option to run in strict parsing mode, such that all errors are thrown rather than recovered.
2. The error thrown contains the location of the error (line, column) as opposed to the replacement node that is passed to `onParseError()`

I'm happy to help implement this if you're open to it.

**Comment by nzakas** [comment]
Sorry, I just realized that `onParseError()` passes a `SyntaxError` object as the first argument that contains all of the relevant information. I can use that to throw an error in ESLint.

I'd still like to see if there's a solution for `a {`. I understand that, by the spec, you are to close the block and consider it valid. However, from a linter's point of view, we'd still like to be able to flag this as an error.

**Comment by nzakas** [comment]
I put together a PR for this with the changes we're looking for:
https://github.com/csstree/csstree/pull/305

**Comment by lahmatiy** [comment]
I don’t believe introducing a new parser mode is necessary for the described goals. Specifically, to create a rule that validates bracket matching, it’s possible to achieve this solely on the token level:  

- Check for the presence of `<bad-string>` and `<bad-url>` tokens and issue warnings for them.  
- For round, square, and curly brackets, validating their balance is sufficient. The parser (and the `TokenStream` it inherits from) already tracks a balance vector (`balance` property), which can be leveraged for this purpose.  

### How `balance` works  

For every non-closing token (including opening ones), the `balance` vector stores the index of the closing token corresponding to the last opening token (for opening tokens, this is the token itself). If there is no closing token, the index points to the end of the `balance` vector (since `balance` is TypedArray, a `tokenCount` should be used instead of `balance.length`). For closing tokens, the `balance` vector stores the index of the corresponding opening token. If no pair exists, the index behaves as it would for any unmatched token, pointing to the closing token of the last opened bracket or the end of the vector if no pair exists.  

In CSS, brackets must be balanced. For instance, in the case of `"(])"`, the square bracket would be unmatched (its index in `balance` would be `>= balance.length`). In the case of `"([)"`, all three brackets would remain unmatched, with their indices in `balance` being `>= balance.length`.  

This allows identifying problematic tokens based on the following rules:  

- For `<function-token>`, `<(-token>`, `<[-token>`, and `<{-token>`, the value of `balance[tokenIndex]` must be `< tokenCount`. Otherwise, the token lacks a closing pair, and a warning can be issued.  
- For `<)-token>`, `<]-token>`, and `<}-token>`, the condition `balance[tokenIndex] < tokenIndex` must hold. Otherwise, the token lacks an opening pair, and a warning can be issued.

A prototype of the solution:

```js
import { tokenize, TokenStream, tokenTypes } from 'css-tree';
import { OffsetToLocation } from 'css-tree/lib/tokenizer/OffsetToLocation.js'; // Looks like I forgot to add this class to the export, needs to be fixed

const css = '(]"bad\n)';
const tokenStream = new TokenStream(css, tokenize);
const offsetToLocation = new OffsetToLocation();
const emitWarning = (start, end, text) => {
    console.warn(text, offsetToLocation.getLocationRange(start, end));
};

offsetToLocation.setSource(css); // Looks ugly, needs to be fixed so that it can be set in the constructor
tokenStream.forEachToken((type, start, end, index) => {
    switch (type) {
        case tokenTypes.BadString:
            emitWarning(start, end, 'Bad string');
            break;

        case tokenTypes.BadUrl:
            emitWarning(start, end, 'Bad url');
            break;

        case tokenTypes.Function:
        case tokenTypes.LeftParenthesis:
        case tokenTypes.LeftSquareBracket:
        case tokenTypes.LeftCurlyBracket:
            if (tokenStream.balance[index] >= tokenStream.tokenCount) {
                emitWarning(start, end, 'No closing pair');
            }
            break;

        case tokenTypes.RightParenthesis:
        case tokenTypes.RightSquareBracket:
        case tokenTypes.RightCurlyBracket:
            if (tokenStream.balance[index] >= index) {
                emitWarning(start, end, 'No opening pair');
            }
            break;
    }
});
```

### A balance issue in the parser  

While revisiting how `balance` works and experimenting with various cases, I discovered that the parser does not correctly maintain bracket balance. For example, if an error occurs while parsing (e.g., a declaration value), the parser skips content until a general rule for ending the construct is met (e.g., `;` or `}`). This behavior is incorrect. In such cases, the parser should rely on bracket balance as described in the specification and implemented by browsers.  

Fortunately, these issues only occur when there’s an actual CSS error, which likely explains why the problem went unnoticed for so long. However, this leads to incorrect parsing of CSS that should otherwise fail (i.e., be skipped). While this leniency might be beneficial for tools in some cases, it results in unpredictable outcomes. It’s better to stay aligned with specifications and browser behavior. Addressing this is part of my plans.  

### On additional parser modes  

Adding new parser modes significantly complicates the logic. CSS does not have a concept of strict vs. non-strict parsers, and introducing such modes seems unnecessary. Additionally, the parser already has options like `parseAtrulePrelude`, `parseRulePrelude`, `parseValue`, and `parseCustomProperty`, which disable parsing parts of CSS. In these cases, no rules are checked in the skipped fragments (tokens are simply concatenated into a single value), so "strict" rules won’t apply.  

Moreover, the parser is moving toward greater tolerance to ensure errors during parsing don’t result in a complete loss of parsed structure (parts of the AST). This allows tools and developers to resolve issues without needing to implement their own parsing logic, safely obtaining a structure instead of raw strings. Errors during parsing currently fall back to treating content as raw or switching to alternative parsing rules. The latter is applied in cases like `MediaQuery` parsing and should also apply to `CSS Nesting`. Initially, CSS is parsed as a declaration (`Declaration`), and if an error occurs, it’s re-parsed as a rule (`Rule`). Here, the "error" doesn’t mean a genuine CSS authoring mistake but simply that the fragment couldn’t be parsed as a `Declaration`. This behavior is by design (see [this](https://drafts.csswg.org/css-syntax/#consume-a-blocks-contents), the "anything else" point, and the implementation note).  

If a new "parser mode" introduces additional errors, it would result in different semantics for the same fragments across modes. This would significantly complicate the logic and introduce inconsistencies.  

Leveraging the parser’s error-handling mechanics to extract additional information is not an optimal approach, as such changes alter the parsing logic defined by specifications.

**Comment by lahmatiy** [comment]
I noticed that everything in the example above can be imported from `css-tree/tokenizer`:

```js
import { tokenize, TokenStream, tokenTypes, OffsetToLocation } from 'css-tree/tokenizer';
```

However, I still believe adding `OffsetToLocation` to the main entry point would be beneficial.

**Comment by nzakas** [comment]
Thanks for the detailed explanation. Let me dig into this and see if I can solve our use case following your example.

**Comment by lahmatiy** [comment]
The issue with balance turned out to be slightly broader — the `balance` vector itself was calculated incorrectly. I’ve fixed it, and now everything works as described in my previous comment. The changes have been published in version `3.1.0`, see detail in [release notes](https://github.com/csstree/csstree/releases/tag/v3.1.0) or change log.

I’d also like to highlight something about working with CSS in general. Unlike most programming languages, CSS is token-centric. Almost everything (parsing, value substitution, syntax matching, serialization, etc.) is handled at the token level rather than the AST. This isn’t specific to CSSTree, it’s how the language is designed and operates. Once you embrace this approach, many tasks become simpler. The AST is primarily useful for grouping tokens and defining hierarchy; for everything else, tokens are the main focus.

**Comment by nzakas** [comment]
Maybe I'm missing something, but it seems like with this solution, I'd need to do two runs over the source code:

1. Parse into the AST
2. Re-run through my own `TokenStream` to catch additional errors

I suppose this is the same issue as: https://github.com/csstree/csstree/issues/205

> The AST is primarily useful for grouping tokens and defining hierarchy; for everything else, tokens are the main focus.

In ESLint, all rules are written against an AST, so while it may be secondary to how CSS is parsed, it is primary for the user experience in ESLint.

**Comment by lahmatiy** [comment]
> I suppose this is the same issue as: https://github.com/csstree/csstree/issues/205  

Yes, exactly. Additionally, we need to provide a way to access `balance` in `onToken`, at the very least.  

> In ESLint, all rules are written against an AST, so while it may be secondary to how CSS is parsed, it is primary for the user experience in ESLint.

I understand, and that’s precisely why I highlighted this aspect. In its early development, CSSTree was also AST-centric, and solving problems was quite challenging because CSS specs and the language itself don’t inherently consider ASTs. In CSS, there are essentially two key projections: tokens and the CSS Object Model (CSSOM).  

Tokens represent a very low-level projection, and most specifications express and operate on them. The CSSOM, on the other hand, is a very high-level abstraction that can sometimes feel too detached from the original CSS source. The AST sits in the middle, but neither the specs nor browsers utilize it (browsers parse CSS directly into CSSOM). As a result, AST-based solutions often need to be invented, leading to suboptimal implementations that fail to cover all cases and time to time break with new CSS spec updates.

At some point, CSSTree shifted its focus to tokens, and this simplification made many problems much easier. Once you embrace tokens as the foundation for working with CSS, everything becomes more straightforward. Tokens are a powerful and essential tool in CSS, setting it apart from other languages.

This shift doesn’t eliminate the need for an AST, but the AST becomes secondary. My suggestion is to consider this approach and potentially add an API that allows direct interaction with tokens alongside the AST. This would provide greater flexibility and robustness when working with CSS.

**Comment by nzakas** [comment]
> This shift doesn’t eliminate the need for an AST, but the AST becomes secondary. My suggestion is to consider this approach and potentially add an API that allows direct interaction with tokens alongside the AST. This would provide greater flexibility and robustness when working with CSS.

ESLint already has the ability to access tokens alongside an AST, we do that already for JavaScript. The problem here is that there's no way to get those tokens without re-tokenizing the input.

**Comment by nzakas** [comment]
Closing as we can most of this now with the onToken() callback.
